crates/adapters/src/format/json/output.rs

```diff
@@ -96,7 +98,15 @@ impl Encoder for JsonEncoder {
     fn encode(&mut self, batches: &[Arc<dyn SerBatch>]) -> AnyResult<()> {
         let mut buffer = take(&mut self.buffer);
 
+        // Reserve one extra byte for the closing bracket `]`.
+        let max_buffer_size = if self.config.array {
+            self.max_buffer_size - 1
+        } else {
+            self.max_buffer_size
+        };
+
         let mut num_records = 0;
+        let mut buffer_full = false;
```

Comment:
The `buffer_full` logic seems more complex than necessary because it is unnecessarily declared in an outer scope. If you care, here is one suggested improvement:

```patch
diff --git a/crates/adapters/src/format/json/output.rs b/crates/adapters/src/format/json/output.rs
index b83a6b53..3679bbec 100644
--- a/crates/adapters/src/format/json/output.rs
+++ b/crates/adapters/src/format/json/output.rs
@@ -106,7 +106,6 @@ impl Encoder for JsonEncoder {
         };
 
         let mut num_records = 0;
-        let mut buffer_full = false;
 
         for batch in batches.iter() {
             let mut cursor = batch.cursor();
@@ -142,7 +141,8 @@ impl Encoder for JsonEncoder {
 
                     // Drop the last encoded record if it exceeds max_buffer_size.
                     // The record will be included in the next buffer.
-                    if buffer.len() > max_buffer_size {
+                    let buffer_full = buffer.len() > max_buffer_size;
+                    if buffer_full {
                         if num_records == 0 {
                             // We should be able to fit at least one record in the buffer.
                             bail!("JSON record exceeds maximum buffer size supported by the output transport. Max supported buffer size is {} bytes, but the following record requires {} bytes: '{}'.",
@@ -151,7 +151,6 @@ impl Encoder for JsonEncoder {
                                   serde_json::to_string(cursor.key()).unwrap_or_default());
                         }
                         buffer.truncate(prev_len);
-                        buffer_full = true;
                     } else {
                         if w > 0 {
                             w -= 1;
@@ -173,7 +172,6 @@ impl Encoder for JsonEncoder {
                         self.output_consumer.push_buffer(&buffer);
                         buffer.clear();
                         num_records = 0;
-                        buffer_full = false;
                     }
                 }
 
```